nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ScriptedPartitionRecord.java

```diff
@@ -0,0 +1,232 @@
+                        final RecordSet recordSet = reader.createRecordSet();
+                        final PushBackRecordSet pushBackSet = new PushBackRecordSet(recordSet);
+                        final Map<String, RecordBatchingProcessorFlowFileBuilder> recordSetFlowFileBuilders = new HashMap<>();
+                        final BiFunction<FlowFile, OutputStream, RecordSetWriter> recordSetWriterFactory = (outgoingFlowFile, out) -> {
+                            try {
+                                return writerFactory.createWriter(getLogger(), schema, out, outgoingFlowFile);
+                            } catch (final IOException | SchemaNotFoundException e) {
+                                throw new ProcessException("Could not create RecordSetWriter", e);
+                            }
+                        };
+
+                        int index = 0;
+
+                        // Reading in records and evaluate script
+                        while (pushBackSet.isAnotherRecord()) {
+                            final Record record = pushBackSet.next();
+                            final Object evaluatedValue = evaluator.evaluate(record, index++);
+                            getLogger().debug("Evaluated scripted against {} (index {}), producing result of {}", record, index - 1, evaluatedValue);
+
+                            if (evaluatedValue != null && evaluatedValue instanceof String) {
+                                final String partition = (String) evaluatedValue;
+
+                                if (!recordSetFlowFileBuilders.containsKey(partition)) {
+                                    recordSetFlowFileBuilders.put(partition, new RecordBatchingProcessorFlowFileBuilder(incomingFlowFile, session, recordSetWriterFactory));
+                                }
+
+                                final int recordCount = recordSetFlowFileBuilders.get(partition).addRecord(record);
+                                session.adjustCounter("Record Processed", recordCount, false);
+
+                            } else {
+                                throw new ProcessException("Script returned a value of " + evaluatedValue
+                                        + " but this Processor requires that the object returned by an instance of String");
+                            }
+                        }
+
+                        // Sending outgoing flow files
+                        int fragmentIndex = 1;
+
+                        for (final Map.Entry<String, RecordBatchingProcessorFlowFileBuilder> entry : recordSetFlowFileBuilders.entrySet()) {
+                            final String partitionName = entry.getKey();
+                            final RecordBatchingProcessorFlowFileBuilder builder = entry.getValue();
+
+                            FlowFile outgoingFlowFile = builder.build();
+                            outgoingFlowFile = session.putAttribute(outgoingFlowFile, "partition", partitionName);
```

Comment:
Calls to `session.putAttribute` are expensive. They create a full copy of the HashMap that holds attributes. This method should not be used unless only adding a single attribute. If adding multiple attributes, should instead create a Map<String, String>, add all necessary attributes, and then use `session.putAllAttributes`